rclpy/rclpy/subscription.py

```diff
@@ -69,6 +75,21 @@ def __init__(
 
         self.event_handlers: QoSEventHandler = event_callbacks.create_event_handlers(
             callback_group, subscription_impl, topic)
+        self.callback_type = Subscription.CallbackType.MessageOnly
+        try:
+            inspect.signature(self.callback).bind(object())
+            return
+        except TypeError:
+            pass
+        try:
+            inspect.signature(self.callback).bind(object(), object())
+            self.callback_type = Subscription.CallbackType.WithMessageInfo
+            return
+        except TypeError:
+            pass
+        raise RuntimeError(
```

Comment:
I think it would be good to add a positive and negative test for this new logic.

Reply:
See https://github.com/ros2/rclpy/pull/922/commits/74afcb34a36aedae36720a569194f6ad0b515745